graph_nets/demos_tf2/shortest_path.ipynb

```diff
@@ -840,17 +862,36 @@
    },
    "outputs": [],
    "source": [
-    "#@title Reset session  { form-width: \"30%\" }\n",
-    "\n",
-    "# This cell resets the Tensorflow session, but keeps the same computational\n",
-    "# graph.\n",
+    "#@title Compiling update_step with `tf_function`  { form-width: \"30%\" }\n",
+    "\"\"\"To recover the speed of TensorFlow 1, we need to use `tf.function` to\n",
+    "compile the update_step into a graph. However, using `tf.function` naively\n",
+    "may futher reduce performance as the number of nodes/edges in the batch can\n",
+    "change across batches, tensorflow will repeatedly trace the function multiple\n",
+    "times for each unique shape of the input tensors.\n",
+    "Instead, we obtain an explicit signature for each input argument to\n",
+    "`update_step` using `utils_tf.specs_from_graphs_tuple` that sets `None` sizes\n",
+    "for variable length axes in graph fields, preventing `tf.function` from having\n",
+    "to trace the function repeatedly.\"\"\"\n",
+    "def update_step(inputs_tr, targets_tr):\n",
+    "    with tf.GradientTape() as tape:\n",
+    "        output_ops_tr = model(inputs_tr, num_processing_steps_tr)\n",
+    "        # Loss.\n",
+    "        loss_tr = create_loss_ops(targets_tr,output_ops_tr)\n",
+    "        loss_tr = tf.math.reduce_sum(loss_tr) / num_processing_steps_tr\n",
+    "\n",
+    "    gradients = tape.gradient(loss_tr, model.trainable_variables)\n",
+    "    optimizer.apply(gradients, model.trainable_variables)\n",
+    "    return  output_ops_tr,loss_tr\n",
+    "\n",
+    "# Get the input signature for that function by obtaining the specs\n",
+    "input_signature = [\n",
+    "  utils_tf.specs_from_graphs_tuple(input_ph),\n",
```

Comment:
it should be 4 spaces in this case, as this is not a nested context, but a line continuation.